File: grpc_api/client_web/src/common/type_value.js
/**
 * @enum {number}
 */
export const DataType = Object.freeze({
    NULL: 0,
    I8: 1,
    I16: 2,
    I32: 3,
    I64: 4,
    I128: 5,
    U8: 6,
    U16: 7,
    U32: 8,
    U64: 9,
    U128: 10,
    F32: 12,
    F64: 13,
    BOOL: 15,
    CHAR: 16,
    STRING: 17,
    BYTES: 18
});

/**
 * @param {number|string} type
 * @returns {number}
 */
export function set_data_type(type) {
    if (typeof type === "number") {
        if (type >= 0 && type <= 18) {
            return type;
        }
    }
    else if (typeof type === "string") {
        switch (type.toUpperCase()) {
            case "I8": return DataType.I8;
            case "I16": return DataType.I16;
            case "I32": return DataType.I32;
            case "I64": return DataType.I64;
            case "I128": return DataType.I128;
            case "U8": return DataType.U8;
            case "U16": return DataType.U16;
            case "U32": return DataType.U32;
            case "U64": return DataType.U64;
            case "U128": return DataType.U128;
            case "F32": return DataType.F32;
            case "F64": return DataType.F64;
            case "BOOL": return DataType.BOOL;
            case "CHAR": return DataType.CHAR;
            case "STRING": return DataType.STRING;
            case "BYTES": return DataType.BYTES;
        }
    }
    return DataType.NULL;
}

/**
 * @param {number} type 
 * @returns {string}
 */
export function get_data_type(type) {
    switch (type) {
        case DataType.I8: return "I8";
        case DataType.I16: return "I16";
        case DataType.I32: return "I32";
        case DataType.I64: return "I64";
        case DataType.I128: return "I128";
        case DataType.U8: return "U8";
        case DataType.U16: return "U16";
        case DataType.U32: return "U32";
        case DataType.U64: return "U64";
        case DataType.U128: return "U128";
        case DataType.F32: return "F32";
        case DataType.F64: return "F64";
        case DataType.BOOL: return "BOOL";
        case DataType.CHAR: return "CHAR";
        case DataType.STRING: return "STRING";
        case DataType.BYTES: return "BYTES";
    }
    return "NULL";
}

/**
 * @param {string} base64 
 * @returns {ArrayBufferLike}
 */
function base64_to_array_buffer(base64) {
    let binaryString = atob(base64);
    let bytes = new Uint8Array(binaryString.length);
    for (let i=0; i<binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes.buffer;
}

/**
 * @param {ArrayBufferLike} buffer 
 * @returns {string}
 */
function array_buffer_to_base64(buffer) {
    let bytes = new Uint8Array(buffer);
    let binaryString = String.fromCharCode.apply(null, bytes);
    return btoa(binaryString);
}

/**
 * @param {string|ArrayBufferLike} base64 
 * @param {number} type 
 * @returns {number|bigint|string|Uint8Array|boolean|null}
 */
export function unpack_data(base64, type) {
    const buffer = base64_to_array_buffer(base64);
    const array = new Uint8Array(buffer);
    const view = new DataView(buffer);
    switch (type) {
        case DataType.I8: 
            if (view.byteLength >= 1) return view.getInt8(view.byteLength - 1);
        case DataType.I16: 
            if (view.byteLength >= 2) return view.getInt16(view.byteLength - 2);
        case DataType.I32: 
            if (view.byteLength >= 4) return view.getInt32(view.byteLength - 4);
        case DataType.I64: 
            if (view.byteLength >= 8) return view.getBigInt64(view.byteLength - 8);
        case DataType.I128: 
            if (view.byteLength >= 8) return view.getBigInt64(view.byteLength - 8);
        case DataType.U8: 
            if (view.byteLength >= 1) return view.getUint8(view.byteLength - 1);
        case DataType.U16: 
            if (view.byteLength >= 2) return view.getUint16(view.byteLength - 2);
        case DataType.U32: 
            if (view.byteLength >= 4) return view.getUint32(view.byteLength - 4);
        case DataType.U64: 
            if (view.byteLength >= 8) return view.getBigUint64(view.byteLength - 8);
        case DataType.U128: 
            if (view.byteLength >= 8) return view.getBigUint64(view.byteLength - 8);
        case DataType.F32:
            if (view.byteLength >= 4) return view.getFloat32(view.byteLength - 4);
        case DataType.F64:
            if (view.byteLength >= 8) return view.getFloat64(view.byteLength - 8);
        case DataType.BOOL:
            for (const byte of array) if (byte) return true;
            return false;
        case DataType.CHAR:
            if (view.byteLength >= 1) return String.fromCharCode(view.getUint8());
        case DataType.STRING:
            return new TextDecoder("utf-8").decode(array);
        case DataType.BYTES:
            return array;
    }
    return null;
}

/**
 * @param {string} base64 
 * @param {number[]} types 
 * @returns {(number|bigint|string|Uint8Array|boolean|null)[]}
 */
export function unpack_data_array(base64, types) {
    const buffer = base64_to_array_buffer(base64);
    let index = 0;
    let values = [];
    for (const type of types) {
        let length = 0;
        if (type == DataType.I8 || type == DataType.U8 || type == DataType.CHAR || type == DataType.BOOL) {
            length = 1;
        }
        else if (type == DataType.I16 || type == DataType.U16) {
            length = 2;
        }
        else if (type == DataType.I32 || type == DataType.U32 || type == DataType.F32) {
            length = 4;
        }
        else if (type == DataType.I64 || type == DataType.U64 || type == DataType.F64) {
            length = 8;
        }
        else if (type == DataType.I128 || type == DataType.U128) {
            length = 16;
        }
        else if (type == DataType.STRING || type == DataType.BYTES) {
            length = 1;
            if (index < buffer.byteLength) {
                const view = new DataView(buffer.slice(index));
                length = view.getUint8();
                index += 1;
            }
        }
        if (index + length > buffer.byteLength) break;
        const value = unpack_data(array_buffer_to_base64(buffer.slice(index, index + length)), type);
        values.push(value);
        index += length;
    }
    return values;
}

/**
 * @param {number|bigint|string|Uint8Array|boolean} value
 */
export function pack_type(value) {
    if (typeof value == "number") {
        if (Number.isInteger(value)) {
            return DataType.I32;
        }
        else {
            return DataType.F64;
        }
    }
    else if (typeof value == "bigint") {
        return DataType.I64;
    }
    else if (typeof value == "string") {
        if (value.length == 1) {
            return DataType.CHAR;
        }
        else {
            return DataType.STRING;
        }
    }
    else if (value instanceof Uint8Array) {
        return DataType.BYTES;
    }
    else if (typeof value == "boolean") {
        return DataType.BOOL;
    }
    return DataType.NULL;
}

/**
 * @param {number|bigint|string|Uint8Array|boolean} value
 */
function pack(value) {
    if (typeof value == "number") {
        if (Number.isInteger(value)) {
            const buffer = new ArrayBuffer(4);
            const view = new DataView(buffer);
            view.setInt32(0, value);
            return view.buffer;
        } else {
            const buffer = new ArrayBuffer(8);
            const view = new DataView(buffer);
            view.setFloat64(0, value);
            return view.buffer;
        }
    }
    else if (typeof value == "bigint") {
        const buffer = new ArrayBuffer(8);
        const view = new DataView(buffer);
        view.setBigInt64(0, value);
        return view.buffer;
    }
    else if (typeof value == "string") {
        const array = new Uint8Array(value.length);
        array.set(new TextEncoder("utf-8").encode(value));
        return array.buffer;
    }
    else if (value instanceof Uint8Array) {
        return value.buffer;
    }
    else if (typeof value == "boolean") {
        let array = new Uint8Array([0]);
        if (value) array = new Uint8Array([1]);
        return array.buffer;
    }
    return new ArrayBuffer(0);
}

/**
 * @param {number|bigint|string|Uint8Array|boolean} value
 */
export function pack_data(value) {
    return array_buffer_to_base64(pack(value));
}

/**
 * @param {(number|bigint|string|Uint8Array|boolean)[]} values
 */
export function pack_data_array(values) {
    if (values === undefined) {
        return "";
    }
    let arrays = new Uint8Array();
    for (const value of values) {
        let data_buffer = pack(value);
        if ((typeof value == "string" && value.length != 1) || value instanceof Uint8Array) {
            const len = new Uint8Array([value.length % 256]);
            const combine = new Uint8Array(arrays.byteLength + 1);
            combine.set(arrays);
            combine.set(len, arrays.byteLength);
            arrays = combine;
        }
        let array = new Uint8Array(data_buffer);
        let combine = new Uint8Array(arrays.byteLength + array.byteLength);
        combine.set(arrays);
        combine.set(array, arrays.byteLength);
        arrays = combine;
    }
    return array_buffer_to_base64(arrays.buffer);
}

File: grpc_api/client_python/src/bbthings_grpc/common/type_value.py
from enum import Enum
from typing import Union, List
from struct import pack, unpack


class DataType(Enum):
    NULL = 0
    I8 = 1
    I16 = 2
    I32 = 3
    I64 = 4
    I128 = 5
    U8 = 6
    U16 = 7
    U32 = 8
    U64 = 9
    U128 = 10
    F32 = 12
    F64 = 13
    BOOL = 15
    CHAR = 16
    STRING = 17
    BYTES = 18


def pack_type(value: Union[int, float, str, bytes, bool, None]):
    if isinstance(value, bool):
        return DataType.BOOL.value
    elif isinstance(value, int):
        return DataType.I64.value
    elif isinstance(value, float):
        return DataType.F64.value
    elif isinstance(value, str):
        return DataType.CHAR.value if len(value) == 1 else DataType.STRING.value
    elif isinstance(value, bytes):
        return DataType.BYTES.value
    else:
        return DataType.NULL.value

def pack_data(value: Union[int, float, str, bytes, bool, None]) -> bytes:
    if type(value) == int:
        mod = abs(value) % 9223372036854775808
        if value < 0: mod = -mod
        return pack('>q', mod)
    elif type(value) == float:
        return pack('>d', value)
    elif type(value) == str:
        return bytes(value, 'utf-8')
    elif type(value) == bytes:
        return value
    elif type(value) == bool:
        if value: return b'\x01'
        else: return b'\x00'
    else:
        return bytes()

def pack_data_array(values: List[Union[int, float, str, bool, None]]) -> bytes:
    binary = bytes()
    for value in values:
        binary_value = pack_data(value)
        if type(value) == bytes or (type(value) == str and len(value) != 1):
            binary = binary + bytes((len(binary_value) % 256,)) # insert length at first element
        binary = binary + binary_value
    return binary

def pack_data_type(value: Union[int, float, str, bool, None], type: DataType):
    if type == DataType.I8:
        mod = abs(int(value)) % 128
        if value < 0: mod = -mod
        return pack('b', mod)
    elif type == DataType.I16:
        mod = abs(int(value)) % 32768
        if value < 0: mod = -mod
        return pack('>h', mod)
    elif type == DataType.I32:
        mod = abs(int(value)) % 2147483684
        if value < 0: mod = -mod
        return pack('>l', mod)
    elif type == DataType.I64:
        mod = abs(int(value)) % 9223372036854775808
        if value < 0: mod = -mod
        return pack('>q', mod)
    elif type == DataType.U8:
        mod = int(value) % 256
        return pack('B', mod)
    elif type == DataType.U16:
        mod = int(value) % 65536
        return pack('>H', mod)
    elif type == DataType.U32:
        mod = int(value) % 4294967296
        return pack('>L', mod)
    elif type == DataType.U64:
        mod = int(value) % 18446744073709551616
        return pack('>Q', mod)
    elif type == DataType.F32:
        return pack('>f', float(value))
    elif type == DataType.F64:
        return pack('>d', float(value))
    elif type == DataType.BOOL:
        if value: return b'\x01'
        else: return b'\x00'
    elif type == DataType.CHAR:
        return bytes(value, 'utf-8')[:1]
    elif type == DataType.STRING:
        return bytes(value, 'utf-8')
    elif type == DataType.BYTES:
        return bytes(value)
    else:
        return bytes()

def pack_data_type_array(values: List[Union[int, float, str, bool, None]], types: List[DataType]) -> bytes:
    binary = bytes()
    for i, type in enumerate(types):
        binary_value = pack_data_type(values[i], type)
        if type == DataType.STRING or type == DataType.BYTES:
            binary = binary + bytes((len(binary_value) % 256,)) # insert length at first element
        binary = binary + binary_value
    return binary

def unpack_data(binary: bytes, type: DataType) -> Union[int, float, str, bool, None]:
    if type == DataType.I8:
        return unpack('b', binary[-1:])[0]
    elif type == DataType.I16:
        return unpack('>h', binary[-2:])[0]
    elif type == DataType.I32:
        return unpack('>l', binary[-4:])[0]
    elif type == DataType.I64:
        return unpack('>q', binary[-8:])[0]
    elif type == DataType.I128:
        return unpack('>q', binary[-16:])[0]
    elif type == DataType.U8:
        return unpack('B', binary[-1:])[0]
    elif type == DataType.U16:
        return unpack('>H', binary[-2:])[0]
    elif type == DataType.U32:
        return unpack('>L', binary[-4:])[0]
    elif type == DataType.U64:
        return unpack('>Q', binary[-8:])[0]
    elif type == DataType.U128:
        return unpack('>Q', binary[-16:])[0]
    elif type == DataType.F32:
        return unpack('>f', binary[-4:])[0]
    elif type == DataType.F64:
        return unpack('>d', binary[-8:])[0]
    elif type == DataType.BOOL:
        for byte in binary:
            if byte != 0: return True
        return False
    elif type == DataType.CHAR:
        return str(binary[:1], 'utf-8')
    elif type == DataType.STRING:
        return str(binary, 'utf-8')
    elif type == DataType.BYTES:
        return binary
    else:
        return None

def unpack_data_array(binary: bytes, types: List[DataType]) -> List[Union[int, float, str, bool, None]]:
    index = 0
    values = []
    for ty in types:
        size = 0
        if ty == DataType.I8 or ty == DataType.U8 or ty == DataType.CHAR or ty == DataType.BOOL:
            size = 1
        elif ty == DataType.I16 or ty == DataType.U16:
            size = 2
        elif ty == DataType.I32 or ty == DataType.U32 or ty == DataType.F32:
            size = 4
        elif ty == DataType.I64 or ty == DataType.U64 or ty == DataType.F64:
            size = 8
        elif ty == DataType.I128 or ty == DataType.U128:
            size = 16
        elif ty == DataType.STRING or ty == DataType.BYTES:
            if index < len(binary): 
                size = int(binary[index]) # first element is the length
                index += 1 # skip first element
            else: size = 1
        if index + size > len(binary): break
        value = unpack_data(binary[index:index + size], ty)
        values.append(value)
        index += size
    return values
